### PR TITLE
Handicap / resource transfer

### LINKED ISSUE #3983: Handicap
**Describe your proposal**
Implement OH3 feature "Handicap".

Absolutely low priority. Rather as documentation of the difference between OH3 and VCMI.

**Additional context**
grafik

### PR DESCRIPTION
**Pull request by Laserlicht**
Todo:
- [x] Correct implementation of % income calculation

Fixes #3983 (with other implementation)

income/growth: if you select below 100%: Mines will still create at least 1 resource. And in the town there is still a growth of at least 1 creature.
To make them not completely useless.

Current status: 
grafik

### CODE REVIEW

**Comment by dydzio0614** [comment]
won't check how this works right now, but make sure positive values add resource and negative values subtract them and that negative value is properly capped (if not in gui then big negative value should cap resources at 0 on server side)

**Comment by Laserlicht** [comment]
> won't check how this works right now, but make sure positive values add resource and negative values subtract them and that negative value is properly capped (if not in gui then big negative value should cap resources at 0 on server side)

Yes, resources will be added. Pos and negative are possible (except for incoming %)
In the first test it was capped correctly to 0. But may have to retest.

**Comment by Laserlicht** [comment]
Also supports de-/increasing growth in town now:
grafik

File: lib/mapObjects/MiscObjects.cpp [hunk 1]
```diff
@@ -195,6 +196,15 @@ ui32 CGMine::defaultResProduction() const
 	}
 }
 
+ui32 CGMine::getProducedQuantity() const
+{
+	auto * playerSettings = cb->getPlayerSettings(getOwner());
+	auto ret = producedQuantity * playerSettings->handicap.percentIncome / 100;
+	if(ret == 0 && producedQuantity > 0) // create at least 1 resource
+		ret = 1;
+	return ret;
```
Comment (IvanSavenko) [inline lib/mapObjects/MiscObjects.cpp:205]: Maybe always round up instead?
```suggestion
	// always round up income - we don't want mines to always produce zero if handicap in use
	return (producedQuantity * playerSettings->handicap.percentIncome + 99) / 100;
```

File: lib/mapObjects/CGTownInstance.cpp [hunk 2]
```diff
@@ -1257,7 +1269,10 @@ int GrowthInfo::totalGrowth() const
 	for(const Entry &entry : entries)
 		ret += entry.count;
 
-	return ret;
+	auto retCalc = ret * percent / 100;
+	if(retCalc == 0 && ret > 0) //generate at least one
+		retCalc = 1;
+	return retCalc;
```
Comment (IvanSavenko) [inline lib/mapObjects/CGTownInstance.cpp:1275]: Same question about rounding up

File: client/lobby/OptionsTab.cpp [hunk 3]
```diff
@@ -923,6 +1031,45 @@ OptionsTab::PlayerOptionsEntry::PlayerOptionsEntry(const PlayerSettings & S, con
 	}
 	labelWhoCanPlay = std::make_shared<CMultiLineLabel>(Rect(6, 23, 45, (int)graphics->fonts[EFonts::FONT_TINY]->getLineHeight()*2), EFonts::FONT_TINY, ETextAlignment::CENTER, Colors::WHITE, CGI->generaltexth->arraytxt[206 + whoCanPlay]);
 
+	labelHandicap = std::make_shared<CMultiLineLabel>(Rect(57, 24, 47, (int)graphics->fonts[EFonts::FONT_TINY]->getLineHeight()*2), EFonts::FONT_TINY, ETextAlignment::CENTER, Colors::WHITE, s->handicap.startBonus.empty() && s->handicap.percentIncome == 100 && s->handicap.percentGrowth == 100 ? CGI->generaltexth->arraytxt[210] : MetaString::createFromTextID("vcmi.lobby.handicap").toString());
```
Comment (IvanSavenko) [inline client/lobby/OptionsTab.cpp:1034]: Simplify this line perhaps? Compute some properties into a separate variable in a separate line to reduce length of this line.

File: client/lobby/OptionsTab.cpp [hunk 4]
```diff
@@ -793,6 +794,113 @@ void OptionsTab::SelectionWindow::showPopupWindow(const Point & cursorPosition)
...
+		close();
+	}, EShortcut::GLOBAL_RETURN));
+
+	updateShadow();
+	center();
+}
+
+bool OptionsTab::HandicapWindow::receiveEvent(const Point & position, int eventType) const
+{
+	return true;  // capture click also outside of window
```
Comment (IvanSavenko) [inline client/lobby/OptionsTab.cpp:895]: Can you instead add comment *why* we need it, and not what this code does?

File: client/lobby/OptionsTab.cpp [hunk 5]
```diff
@@ -793,6 +794,113 @@ void OptionsTab::SelectionWindow::showPopupWindow(const Point & cursorPosition)
...
+				textinputs[player][resource]->setText(tmp.length() == 0 ? "0" : (negative ? "-" : "") + std::to_string(stoi(tmp)));
+			});
+			textinputs[player][resource]->setPopupCallback([isIncome, isGrowth](){
+				if(isIncome)
+					CRClickPopup::createAndPush(CGI->generaltexth->translate("vcmi.lobby.handicap.income"));
+				else if(isGrowth)
+					CRClickPopup::createAndPush(CGI->generaltexth->translate("vcmi.lobby.handicap.growth"));
+				else
+					CRClickPopup::createAndPush(CGI->generaltexth->translate("vcmi.lobby.handicap.resource"));
+			});
```
Comment (IvanSavenko) [inline client/lobby/OptionsTab.cpp:859]: Perhaps extract these lambdas into a separate method for clarity?

Comment (Laserlicht) [reply client/lobby/OptionsTab.cpp]: added comments for clarity

**Comment by IvanSavenko** [comment]
> Conflicting files
lib/serializer/ESerializationVersion.h 

Please merge with develop

**Pull request merged**